File: donors/orchestration-java/EventRouter.java
package com.aiassistant.core.orchestration;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class EventRouter {
    private static final String TAG = "EventRouter";
    
    private final Context context;
    private final Map<String, List<EventSubscriber>> subscribers;
    private final ExecutorService executorService;
    private final Handler mainHandler;
    
    public EventRouter(Context context) {
        this.context = context.getApplicationContext();
        this.subscribers = new ConcurrentHashMap<>();
        this.executorService = Executors.newCachedThreadPool();
        this.mainHandler = new Handler(Looper.getMainLooper());
    }
    
    public void subscribe(String eventType, EventSubscriber subscriber) {
        subscribers.computeIfAbsent(eventType, k -> new CopyOnWriteArrayList<>()).add(subscriber);
        Log.d(TAG, "Subscriber added for event type: " + eventType);
    }
    
    public void unsubscribe(String eventType, EventSubscriber subscriber) {
        List<EventSubscriber> eventSubscribers = subscribers.get(eventType);
        if (eventSubscribers != null) {
            eventSubscribers.remove(subscriber);
        }
    }
    
    public void publish(OrchestrationEvent event) {
        publish(event, false);
    }
    
    public void publish(OrchestrationEvent event, boolean synchronous) {
        String eventType = event.getEventType();
        List<EventSubscriber> eventSubscribers = subscribers.get(eventType);
        
        if (eventSubscribers == null || eventSubscribers.isEmpty()) {
            Log.d(TAG, "No subscribers for event: " + eventType);
            return;
        }
        
        Log.d(TAG, "Publishing event: " + eventType + " from " + event.getSource() + 
              " to " + eventSubscribers.size() + " subscribers");
        
        if (synchronous) {
            for (EventSubscriber subscriber : eventSubscribers) {
                try {
                    subscriber.onEvent(event);
                } catch (Exception e) {
                    Log.e(TAG, "Error in subscriber for event " + eventType, e);
                }
            }
        } else {
            executorService.execute(() -> {
                for (EventSubscriber subscriber : eventSubscribers) {
                    try {
                        subscriber.onEvent(event);
                    } catch (Exception e) {
                        Log.e(TAG, "Error in subscriber for event " + eventType, e);
                    }
                }
            });
        }
    }
    
    public void publishOnMainThread(OrchestrationEvent event) {
        mainHandler.post(() -> publish(event, true));
    }
    
    public void shutdown() {
        executorService.shutdown();
    }
    
    public interface EventSubscriber {
        void onEvent(OrchestrationEvent event);
    }
}

File: android/app/src/main/kotlin/com/ariaagent/mobile/core/orchestration/EventRouter.kt
// Ported from donors/orchestration-java/EventRouter.java
// Original repo: TITANICBHAI/AI-ASSISTANT-INCOMPLETE
// Changes: coroutine-based fan-out (Dispatchers.Default), MutableSharedFlow-style
// subscriber registry, topic + wildcard subscriptions, no Android Handler/Looper.

package com.ariaagent.mobile.core.orchestration

import android.util.Log
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Event router for the orchestration layer.
 *
 * Each subscriber is registered against an event-type string (or [WILDCARD]
 * for "all events"). [publish] fans out to subscribers on a background
 * coroutine; [publishSync] runs them on the caller's thread for cases that
 * need ordering (state-machine transitions, tests).
 *
 * This is intentionally separate from [com.ariaagent.mobile.core.events.AgentEventBus]:
 * that bus is the agent's UI-facing event firehose; this router carries
 * lifecycle / orchestration events between system components.
 */
class EventRouter {

    /** subscriber callback type: receives the event, runs on Dispatchers.Default. */
    fun interface Subscriber {
        fun onEvent(event: OrchestrationEvent)
    }

    private val subscribers = ConcurrentHashMap<String, CopyOnWriteArrayList<Subscriber>>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    fun subscribe(eventType: String, subscriber: Subscriber) {
        subscribers.computeIfAbsent(eventType) { CopyOnWriteArrayList() }.add(subscriber)
        Log.d(TAG, "Subscriber added for event type: $eventType")
    }

    fun unsubscribe(eventType: String, subscriber: Subscriber) {
        subscribers[eventType]?.remove(subscriber)
    }

    /** Fire-and-forget publish onto Dispatchers.Default. */
    fun publish(event: OrchestrationEvent) {
        scope.launch { dispatch(event) }
    }

    /** Synchronous publish on the caller's thread. */
    fun publishSync(event: OrchestrationEvent) {
        dispatch(event)
    }

    private fun dispatch(event: OrchestrationEvent) {
        val typed = subscribers[event.eventType].orEmpty()
        val wildcard = subscribers[WILDCARD].orEmpty()

        if (typed.isEmpty() && wildcard.isEmpty()) {
            Log.d(TAG, "No subscribers for event: ${event.eventType}")
            return
        }
        Log.d(
            TAG,
            "Publishing event: ${event.eventType} from ${event.source} " +
                "to ${typed.size + wildcard.size} subscriber(s)",
        )
        for (sub in typed) safeDeliver(sub, event)
        for (sub in wildcard) safeDeliver(sub, event)
    }

    private fun safeDeliver(subscriber: Subscriber, event: OrchestrationEvent) {
        try {
            subscriber.onEvent(event)
        } catch (t: Throwable) {
            Log.e(TAG, "Error in subscriber for event ${event.eventType}", t)
        }
    }

    fun shutdown() {
        scope.cancel()
        subscribers.clear()
    }

    companion object {
        private const val TAG = "EventRouter"
        const val WILDCARD = "*"
    }
}
